---
title: 'Effect of treatment with SHAM1 on T-ALL cells expression'
author: 'Ana del Val'
subtitle: '`r params$subtitulo`'
date: '`r format(Sys.Date(),"%e de %B, %Y")`' 
# date: \today  (solo para pdf)
output:
  #pdf_document:
   # keep_tex: no
    #number_sections: yes
    #toc: yes
    #toc_depth: 3
  html_document:
    toc: yes
    toc_depth: 3
params:
  subtitulo: Análisis de datos ómicos
geometry: margin=2cm
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(cache=TRUE)
options(width=90)
```
``

```{r include=FALSE}
library("rmarkdown")
library("knitr")
library("yaml")
library("tinytex")
library("formatR")
library("caTools")
library("bitops")
library("tools")
library("utils")
library("htmltools")
library("evaluate")
library("base64enc")
library("jsonlite")
library("rprojroot")
library("mime")
library("methods")
library("stringr")
library("animation")
library("kableExtra")
library("class")
library("gmodels")
library("caret")
library("vcd")
library("irr")
library("ROCR")
library("animation")
library("BiocManager")
library("Biostrings")
#install.packages("BiocManager")
#BiocManager::install("Biostrings")
library("ggseqlogo")
library("ggplot2")
library("NLP")
library("tm")
library("SnowballC")
library("RColorBrewer")
library("wordcloud")
library("e1071")
library("neuralnet")
library("nnet")
library("RSNNS")
library("rhandsontable")
library("RODBC")
library("shinydashboard")
library("shiny")
library("NeuralNetTools")
library("caret")
library("klaR")
library("kernlab")
library("shiny")
library("dplyr")
library("RSNNS")

citation("Biobase")

#if (!requireNamespace("BiocManager", quietly = TRUE))
#install.packages("ReactomePA")
#BiocManager::install("ReactomePA")

#citation("oligo")
library("Biobase")
library("GEOquery")
library("devtools")
library("roxygen2")
#library("testhat")
library("knitr")
library("plier")
library("rpart")
library("simpleaffy")
#library("tkWidgets")
library("yeastCC")
library("ygs98.db")
library("ygs98cdf")
library("ygs98probe")
library("affy")
library("affyPLM")
library("affyQCReport")
library("annotate")
library("arrayQualityMetrics")
library("Biobase")
library("CMA")
library("cMAP")
library("e1071")
library("gcrma")
library("genefilter")
library("GO.db")
library("golubEsets")
library("gplots")
library("hgu95av2.db")
library("KEGG.db")
library("limma")
library("multtest")
library("oligo")
library("ggrepel")
library("pvca")
library("mogene21sttranscriptcluster.db")
library("ReactomePA")
```


\pagebreak

#Ejercicio 4, Debate 2

The dataset for the exercise is available at the entry Series GSE18198 of the in Gene Expression Omnibus. It consists in the analysis of expression profiles of human T-ALL cell lines treated with DMSO or SAHM1. In short, NOTCH proteins regulate signaling pathways involved in cellular differentiation, proliferation and death. Overactive Notch signaling as been observed in numerous cancers and has been extensively studied in the context of T-cell acute lymphoblastic leukemia (T-ALL) where more than 50% of pateints harbour mutant NOTCH1. Small molecule modulators of these proteins would be important for understanding the role of NOTCH proteins in malignant and normal biological processes. In this stuy, researchers were interested in measuring the global changes in gene expression upon treatment of the human T-ALL cell lines HPB-ALL and KOPT-K1 with either vehicle alone (DMSO) or SAHM1, an alpha-helical hydrocarbon stapled peptide derived from the MAML1 co-activator protein.Therefore, they designed an experiment that consists in triplicate cultures of KOPT-K1 or HPB-ALL cells treated with either DMSO alone or SAHM1 (20 uM) for 24 hours. Total RNA was extracted and hybridized to Affymetrix human U133 plus 2.0 microarrays (three arrays per treatment per cell line for a total of 12 arrays).

1. Describe -and name- the experimental design. Identify the experimental factors and their levels.

2. Write down the design matrix associated with this experimental design.

3. Build the contrast matrix that can be used to answer the following questions:

- Compare the effect of SHAM1 in KOPT-K1 cell line: KOPT-K1 treated with SHAM1 vs KOPT-K1 treated with DMSO (the vehicle)

- The effect of SHAM1 in HPB-ALL cell line: HPB-ALL treated with SHAM1 vs HPB-ALL treated with DMSO.

- The interaction: the differences between the two previous effects.

## 0. Environment preparation, GitHub sync and data preparation

This code is already sync with github repository [link] (https://github.com/AnadelVal/Effect-treatment-SHAM1-on-T-ALL-cells-expression), which is public. 

```{r include=FALSE}
setwd(".")
#getwd()
dir.create("data") #create folder "data"
dir.create("results") #create folder "results"
dir.create("figures") #create folder "figures"
```

Reference [link] (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE18198)

```{r}
getGEOSuppFiles("GSE18198")
#move .tar from downloads to directory. Put the correct file name:
untar("GSE18198_RAW.tar", exdir="data")
cels <- list.files("data/", pattern = "[gz]")
length(cels)
sapply(paste("data", cels, sep="/"), gunzip)
cels
#targets.csv ha sido creada a mano, 12 arrays con estas columnas: FileName, Group, Genotype, Treatment, ShortName
targets <- read.csv2("./data/targets.csv", header = TRUE, sep = ";")
targets
```

## 1. Read data

### Read CEL files
```{r}
celFiles <- list.celfiles("./data", full.names = TRUE)
```

### Read targets
```{r}
my.targets <-read.AnnotatedDataFrame(file.path("./data","targets.csv"),     header = TRUE, row.names = 1, sep=";") 
```

### Read all
```{r}
rawData <- read.celfiles(celFiles, phenoData = my.targets)
#expressionSet is designed to combine several different sources of information into a single convenient structure
my.targets@data$ShortName->rownames(pData(rawData))
colnames(rawData) <-rownames(pData(rawData)) 
head(rawData)
```

## 2. Exploration

### Quality control of raw data
¿The data have enough quality for normalization?
If one array is above a certain threshold defined in the function it is marked with an asterisk as an outlier. When a certain array is marked three times it should be revised carefully.

8 outliers have been detcted, with only 1 star.
```{r}
arrayQualityMetrics(rawData, file=file.path("./results", " .pdf"))
```
The directory is created in results folder. we have to look for a file called index.html, which opens a web page from where we will be able to access a summary of the analysis performed.
_It has not been created there and I had to move it manually, even if adding file.path_

### PCA
```{r}
plotPCA3 <- function (datos, labels, factor, title, scale,colores, size = 1.5, glineas = 0.25) {data <- prcomp(t(datos),scale=scale)
# plot adjustments
dataDf <- data.frame(data$x)
Group <- factor
loads <- round(data$sdev^2/sum(data$sdev^2)*100,1)
# main plot
p1 <- ggplot(dataDf,aes(x=PC1, y=PC2)) +
theme_classic() +
geom_hline(yintercept = 0, color = "gray70") +
geom_vline(xintercept = 0, color = "gray70") +
geom_point(aes(color = Group), alpha = 0.55, size = 3) +
coord_cartesian(xlim = c(min(data$x[,1])-5,max(data$x[,1])+5)) +
scale_fill_discrete(name = "Group")
# avoiding labels superposition
p1 + geom_text_repel(aes(y = PC2 + 0.25, label = labels),segment.size = 0.25, size = size) + 
labs(x = c(paste("PC1",loads[1],"%")),y=c(paste("PC2",loads[2],"%"))) +  
ggtitle(paste("Principal Component Analysis for: ",title,sep=" "))+ 
theme(plot.title = element_text(hjust = 0.5)) +
scale_color_manual(values=colores)
}
```
It shows the scatterplot of the first two principal components performed on the raw data.

First component of the PCA accounts for 89.9% of the total variability of the samples, and as we can observe in the plot, this variability is mainly contributed by the modulator of the protein, as HPB-All are on the right and KPOT-K1 are on the left.
```{r}
plotPCA3(exprs(rawData), labels = targets$ShortName, factor = targets$Group, 
title="Raw data", scale = FALSE, size = 3, 
colores = c("red", "blue", "green", "yellow"))
```

Save image to tiff file in figures folder.
```{r}
tiff("figures/PCA_RawData.tiff", res = 200, width = 4.5, height = 4, units = 'in')
plotPCA3(exprs(rawData), labels = targets$ShortName, factor = targets$Group, 
title="Raw data", scale = FALSE, size = 2, 
colores = c("red", "blue", "green", "yellow"))
dev.off()
```

With a boxplot we visualize the intensity distribution of the arrays.
```{r}
boxplot(rawData, cex.axis=0.5, las=2,  which="all", 
col = c(rep("red", 3), rep("blue", 3), rep("green", 3), rep("yellow", 3)),
main="Distribution of raw intensity values")
```
```{r}
tiff("figures/Intensity_RawData.tiff", res = 200, width =8, height = 5, units = 'in')
boxplot(rawData, cex.axis=0.5, las=2,  which="all", 
col = c(rep("red", 3), rep("blue", 3), rep("green", 3), rep("yellow", 3)),
main="Distribution of raw intensity values")
dev.off()
```
 A light variation of intensity among arrays is observed, but this is the expected for raw data.

### Data normalization

```{r}
eset_rma <- rma(rawData)
```

### Quality control of normalized data

All outliers have disappeared.
```{r}
arrayQualityMetrics(eset_rma, outdir = file.path("./results", "QCDir.Norm"), force=TRUE)
```

```{r}
plotPCA3(exprs(eset_rma), labels = targets$ShortName, factor = targets$Group, 
title="Normalized data", scale = FALSE, size = 3, 
colores = c("red", "blue", "green", "yellow"))
```

```{r}
tiff("figures/Norm/PCA_RawData.tiff", res = 200, width = 4.5, height = 4, units = 'in')
plotPCA3(exprs(eset_rma), labels = targets$ShortName, factor = targets$Group, 
title="Normalized data", scale = FALSE, size = 3, 
colores = c("red", "blue", "green", "yellow"))
dev.off()
```

Now first component accounts for 79.7% of the total variability. Again, it separates samples by modulator.

```{r}
boxplot(eset_rma, cex.axis=0.5, las=2,  which="all", 
col = c(rep("red", 3), rep("blue", 3), rep("green", 3), rep("yellow", 3)),
main="Boxplot for arrays intensity: Normalized Data")
```



```{r}
tiff("figures/Norm/Intensity_RawData.tiff", res = 200, width = 4.5, height = 4, units = 'in')
boxplot(eset_rma, cex.axis=0.5, las=2,  which="all", 
col = c(rep("red", 3), rep("blue", 3), rep("green", 3), rep("yellow", 3)),
main="Boxplot for arrays intensity: Normalized Data")
dev.off()
```

### Batch detection
Time and place-dependent experimental variations
PVCA

```{r}
pData(eset_rma) <- targets
pct_threshold <- 0.6
batch.factors <- c("Genotype", "Treatment")
pvcaObj <- pvcaBatchAssess (eset_rma, batch.factors, pct_threshold)
```

```{r}
bp <- barplot(pvcaObj$dat, xlab = "Effects",
ylab = "Weighted average proportion variance",
ylim= c(0,1.1),col = c("mediumorchid"), las=2,
main="PVCA estimation")

axis(1, at = bp, labels = pvcaObj$label, cex.axis = 0.55, las=2)

values = pvcaObj$dat

new_values = round(values , 3)

text(bp,pvcaObj$dat,labels = new_values, pos=3, cex = 0.5)
```
```{r}
tiff("figures/Batch/pvca.tiff", res = 200, width = 4.5, height = 4, units = 'in')
bp <- barplot(pvcaObj$dat, xlab = "Effects",
ylab = "Weighted average proportion variance",
ylim= c(0,1.1),col = c("mediumorchid"), las=2,
main="PVCA estimation")

axis(1, at = bp, labels = pvcaObj$label, cex.axis = 0.55, las=2)

values = pvcaObj$dat

new_values = round(values , 3)

text(bp,pvcaObj$dat,labels = new_values, pos=3, cex = 0.5)
dev.off()
```

### Detecting most variable genes
```{r}
sds <- apply (exprs(eset_rma), 1, sd)
sdsO<- sort(sds)
plot(1:length(sdsO), sdsO, main="Distribution of variability for all genes",
sub="Vertical lines represent 90% and 95% percentiles",
xlab="Gene index (from least to most variable)", ylab="Standard deviation")
abline(v=length(sds)*c(0.9,0.95))
```

Values of standard deviations allong all samples for all genes ordered from smallest to biggest

```{r}
tiff("figures/Gene variability/gene_variability.tiff", res = 200, width = 4.5, height = 4, units = 'in')
plot(1:length(sdsO), sdsO, main="Distribution of variability for all genes",
sub="Vertical lines represent 90% and 95% percentiles",
xlab="Gene index (from least to most variable)", ylab="Standard deviation")
abline(v=length(sds)*c(0.9,0.95))
dev.off()
```
### Know annotations package
```{r}
fun <- function(GDSDataobj){
    fulltitle <- Meta(GDSDataobj)$title
    title <- strsplit(fulltitle, "\\[|\\]")[[1]][2]
    title <- paste0(gsub("_|-| ", "", tolower(title)), ".db")
    title
}
```
GSE18198 > "hgu133plus2.db"
```{r}
eset <- getGEO("GSE18198")[[1]]
fun(getGEO(annotation(eset)))
```


### Filter least variable genes

```{r}
annotation(eset_rma) <- "hgu133plus2"
filtered <- nsFilter(eset_rma, 
require.entrez = TRUE, remove.dupEntrez = TRUE,
var.filter=TRUE, var.func=IQR, var.cutoff=0.75, 
filterByQuantile=TRUE, feature.exclude = "^AFFX")
```

```{r}
print(filtered$filter.log)
```
```{r}
#eset_rma #54675
eset_filtered <-filtered$eset
eset_filtered #5046 features (genes) left
```
```{r}
dim(exprs(eset_filtered))[1]
```

After filtering, there are `r dim(exprs(eset_filtered))[1]` genes left.

### Save normalized data
```{r}
write.csv(exprs(eset_rma), file="./results/normalized.Data.csv")
write.csv(exprs(eset_filtered), file="./results/normalized.Filtered.Data.csv")
save(eset_rma, eset_filtered, file="./results/normalized.Data.Rda")
```

## 3. Define the experimental setup

Compare gene expression between groups

The Linear Models for Microarrays method, implemented in the limma package Smyth (2005) is used to select differential expressed genes.

### Create the design matrix

The first step for the analysis based on linear models is to create the design matrix. Basically it is a table that describes the allocation of each sample to a group or experimental condition. It has as many rows as samples and as many columns as groups (if only one factor is considered). Each row contains a one in the column of the group to which the sample belongs and a zero in the others.
1 model of 1 factor with 4 levels defined in Targets>Groups > 4 columns.

```{r}
designMat<- model.matrix(~0+Group, pData(eset_filtered))
colnames(designMat) <- c("HPB_ALL.DMSO", "HPB_ALL.SAHM1", "KOPT_K1.DMSO", "KOPT_K1.SAHM1") #se ponen _ en vez de - porque luego hay que restar
```

### Defining comparisons with the Contrasts Matrix

It consists of as many columns as comparisons and as many rows as groups (that is, as columns of the design matrix).

A comparison between groups - called “contrast” - is represented by a “1” and a “-1” in the rows of groups to compare and zeros in the rest. 

3 comparisons > 3 columns in the contrast matrix.

Build the contrast matrix that can be used to answer the following questions:

- Compare the effect of SHAM1 in KOPT-K1 cell line: KOPT-K1 treated with SHAM1 vs KOPT-K1 treated with DMSO (the vehicle)

- The effect of SHAM1 in HPB-ALL cell line: HPB-ALL treated with SHAM1 vs HPB-ALL treated with DMSO.

- The interaction: the differences between the two previous effects.

```{r}
cont.matrix <- makeContrasts (HPB_ALL.DMSOvsHPB_ALL.SAHM1 = HPB_ALL.DMSO-HPB_ALL.SAHM1,KOPT_K1.DMSOvsKOPT_K1.SAHM1 = KOPT_K1.DMSO-KOPT_K1.SAHM1,INT = (HPB_ALL.DMSO-HPB_ALL.SAHM1) - (KOPT_K1.DMSO-KOPT_K1.SAHM1),levels=designMat)
print(cont.matrix)
```

### Model estimation and gene selection

With LIMMA, once the design matrix and the contrasts have been defined, we can proceed to estimate the model, estimate the contrasts and perform the significance tests that will lead to the decision, for each gene and each comparison, if they can be considered differential expressed.

The analysis provides the usual test statistics such as Fold-change t-moderated or adjusted p-values that are used to order the genes from more unless differential expressed.

In order to control the percentage of false positives that may result from high number of contrasts made simultaneously the p-values are adjusted so that we have control over the false positive rate using the Benjamini and Hochberg method Benjamini and Hochberg (1995).

```{r}
fit<-lmFit(eset_filtered, designMat)
fit.main<-contrasts.fit(fit, cont.matrix)
fit.main<-eBayes(fit.main)
class(fit.main)
```

topTable: for a given contrast a list of genes ordered from smallest to biggest p–value which can be considered to be most to least differential expressed.

For Comparison 1:
```{r}
topTab_HPB_ALL.DMSOvsHPB_ALL.SAHM1 <- topTable (fit.main, number=nrow(fit.main), coef="HPB_ALL.DMSOvsHPB_ALL.SAHM1", adjust="fdr") 
head(topTab_HPB_ALL.DMSOvsHPB_ALL.SAHM1)
```

For Comparison 2:
```{r}
topTab_KOPT_K1.DMSOvsKOPT_K1.SAHM1 <- topTable (fit.main, number=nrow(fit.main), coef="KOPT_K1.DMSOvsKOPT_K1.SAHM1", adjust="fdr") 
head(topTab_KOPT_K1.DMSOvsKOPT_K1.SAHM1)
```

For Comparison 3:
```{r}
topTab_INT <- topTable (fit.main, number=nrow(fit.main), coef="INT", adjust="fdr") 
head(topTab_INT)
```

First column of each topTable contains the manufacturer’s (Affymetrix) ID for each probeset. Next step is to guess which gene correspond to each Affymetrix ID. This process is called annotation.
Gene Symbol, the Entrez Gene identifier or the Gene description.

### Gene annotation

```{r}
annotatedTopTable <- function(topTab, anotPackage)
{
topTab <- cbind(PROBEID=rownames(topTab), topTab)
myProbes <- rownames(topTab)
thePackage <- eval(parse(text = anotPackage))
geneAnots <- select(thePackage, myProbes, c("SYMBOL", "ENTREZID", "GENENAME"))
annotatedTopTab<- merge(x=geneAnots, y=topTab, by.x="PROBEID", by.y="PROBEID")
return(annotatedTopTab)
}
```

Annotation tables, one per comparison:

```{r}
topAnnotated_HPB_ALL.DMSOvsHPB_ALL.SAHM1 <- annotatedTopTable(topTab_HPB_ALL.DMSOvsHPB_ALL.SAHM1, anotPackage="hgu133plus2.db")

topAnnotated_KOPT_K1.DMSOvsKOPT_K1.SAHM1 <- annotatedTopTable(topTab_KOPT_K1.DMSOvsKOPT_K1.SAHM1,
anotPackage="hgu133plus2.db")
 
topAnnotated_INT <- annotatedTopTable(topTab_INT,
anotPackage="hgu133plus2.db")

write.csv(topAnnotated_HPB_ALL.DMSOvsHPB_ALL.SAHM1, file="./results/topAnnotated_HPB_ALL.DMSOvsHPB_ALL.SAHM1.csv")

write.csv(topAnnotated_KOPT_K1.DMSOvsKOPT_K1.SAHM1, file="./results/topAnnotated_KOPT_K1.DMSOvsKOPT_K1.SAHM1.csv")

write.csv(topAnnotated_INT, file="./results/topAnnotated_INT.csv")
```
```{r}
short_HPB_ALL.DMSOvsHPB_ALL.SAHM1 <- head(topAnnotated_HPB_ALL.DMSOvsHPB_ALL.SAHM1[1:5,1:4]) #Comparison 1
short_HPB_ALL.DMSOvsHPB_ALL.SAHM1
```

### Visualizating differential expression

1 per comparison. This is for HPB_ALL.DMSOvsHPB_ALL.SAHM1
```{r}
geneSymbols <- select(hgu133plus2.db, rownames(fit.main), c("SYMBOL"))
SYMBOLS<- geneSymbols$SYMBOL
volcanoplot(fit.main, coef=1, highlight=4, names=SYMBOLS, 
main=paste("Differentially expressed genes", colnames(cont.matrix)[1], sep="\n"))
abline(v=c(-1,1))
```

```{r}
tiff("figures/VolcanoPlot.tiff", res = 150, width = 5, height = 5, units = 'in')
volcanoplot(fit.main, coef=1, highlight=4, names=SYMBOLS, 
main=paste("Differentially expressed genes", colnames(cont.matrix)[1], sep="\n"))
abline(v=c(-1,1))
dev.off()
```
The names of the top 4 genes are shown in blue in the figure.

We should do the same with the 2 other comparisons.

### Multiple comparisons
When one selects genes in several comparisons it is usually interesting to know which genes have been selected in each comparison. Sometimes biologically relevant genes will be those that are selected in one of them but not in others. In other occasions he interest will lie in genes that are selected in all comparisons.

This object has as many columns as comparisons and as many rows as genes: 5046x3.
```{r}
res<-decideTests(fit.main, method="separate", adjust.method="fdr", p.value=0.1, lfc=1)
```

Per each gene and comparison a “+1” denotes significantly up-regulated (t-test values >0, FDR < selected cutoff), a “-1” significantly down-regulated (t-test values <0, FDR < selected cutoff) and a “0” non significant difference (FDR > selected cutoff).

```{r}
sum.res.rows<-apply(abs(res),1,sum)
res.selected<-res[sum.res.rows!=0,] 
print(summary(res))
```

This can be visualized in a Venn Diagram.

```{r}
vennDiagram (res.selected[,1:3], cex=0.9)
title("Genes in common between the three comparisons\n Genes selected with FDR < 0.1 and logFC > 1")
```

```{r}
tiff("figures/VennPlot.tiff", res = 150, width = 5.5, height = 5.5, units = 'in')
vennDiagram (res.selected[,1:3], cex=0.9)
title("Genes in common between the three comparisons\n Genes selected with FDR < 0.1 and logFC > 1")
dev.off()
```

Venn diagram showing the genes in common between the three comparisons performed

## 4. Expression profiles visualization: Heatmaps

Genes that have been selected as differential expressed may be visualized using a heatmap. These plots use color palettes to highlight distinct values –here positive (up-regulation) or negative (down-regulation) significantly differential expressions.

Heatmaps can be used to visualize the expression values of differential expressed genes with no specific order, but it is usually preferred to plot them doing a hierarchical clustering on genes (rows) or columns(samples) in order to find groups of genes with common patterns of variation which can eventually be associated to the different groups being compared.

A common option is to select the gens that have been selected in the previous steps, that is the genes that have been called differential expressed in at least one comparison.

```{r}
probesInHeatmap <- rownames(res.selected)
HMdata <- exprs(eset_filtered)[rownames(exprs(eset_filtered)) %in% probesInHeatmap,]

geneSymbols <- select(hgu133plus2.db, rownames(HMdata), c("SYMBOL"))
SYMBOLS<- geneSymbols$SYMBOL
rownames(HMdata) <- SYMBOLS
write.csv(HMdata, file = file.path("./results/data4Heatmap.csv"))
```

(FDR < 0.1 and logFC > 1)
```{r}
my_palette <- colorRampPalette(c("blue", "red"))(n = 299)
 
heatmap.2(HMdata,
Rowv = FALSE,
Colv = FALSE,
main = "Differentially expressed genes \n FDR < 0,1, logFC >=1",
scale = "row",
col = my_palette,
sepcolor = "white",
sepwidth = c(0.05,0.05),
cexRow = 0.5,
cexCol = 0.9,
key = TRUE,
keysize = 1.5,
density.info = "histogram",
ColSideColors = c(rep("red",3),rep("blue",3), rep("green",3), rep("yellow",3)),
tracecol = NULL,
dendrogram = "none",
srtCol = 30)
```
```{r}
tiff("figures/Heatmap1.tiff", res = 150, width = 5.5, height = 5.5, units = 'in')
heatmap.2(HMdata,
Rowv = FALSE,
Colv = FALSE,
main = "Differentially expressed genes \n FDR < 0,1, logFC >=1",
scale = "row",
col = my_palette,
sepcolor = "white",
sepwidth = c(0.05,0.05),
cexRow = 0.5,
cexCol = 0.9,
key = TRUE,
keysize = 1.5,
density.info = "histogram",
ColSideColors = c(rep("red",3),rep("blue",3), rep("green",3), rep("yellow",3)),
tracecol = NULL,
dendrogram = "none",
srtCol = 30)
dev.off()
```


Genes and samples are forced to group by row and column similarity respectively.

```{r}
heatmap.2(HMdata,
Rowv = TRUE,
Colv = TRUE,
dendrogram="both",
main = "Differentially expressed genes \n FDR < 0,1, logFC >=1",
scale = "row",
col = my_palette,
sepcolor = "white",
sepwidth = c(0.05,0.05),
cexRow = 0.5,
cexCol = 0.9,
key = TRUE,
keysize = 1.5,
density.info = "histogram",
ColSideColors = c(rep("red",3),rep("blue",3), rep("green",3), rep("yellow",3)),
tracecol = NULL,
srtCol = 30)
```
```{r}
tiff("figures/Heatmap_grouped.tiff", res = 150, width = 5.5, height = 5.5, units = 'in')
heatmap.2(HMdata,
Rowv = TRUE,
Colv = TRUE,
dendrogram="both",
main = "Differentially expressed genes \n FDR < 0,1, logFC >=1",
scale = "row",
col = my_palette,
sepcolor = "white",
sepwidth = c(0.05,0.05),
cexRow = 0.5,
cexCol = 0.9,
key = TRUE,
keysize = 1.5,
density.info = "histogram",
ColSideColors = c(rep("red",3),rep("blue",3), rep("green",3), rep("yellow",3)),
tracecol = NULL,
srtCol = 30)
dev.off()
```


## 5. Biological Significance of results
Given a list of genes selected for being differential expressed between two conditions, the functions, biological processes or molecular pathways that characterize them appear on this list more frequently than among the rest of the genes analyzed.

ReactomePA Bioconductor package. The analysis is done on the ReactomePA annotation database https://reactome.org/.

Analyses of this type need a minimum number of genes to be reliable, preferably a few hundreds than a few dozens, so it is common to perform a selection less restrictive than with the previous steps. For instance an option is to include all genes with a non-stringent FDR cutoff, such as FDR < 0.15 without filtering by minimum “fold-change”).

```{r}
listOfTables <- list(HPB_ALL.DMSOvsHPB_ALL.SAHM1 = topTab_HPB_ALL.DMSOvsHPB_ALL.SAHM1, 
KOPT_K1.DMSOvsKOPT_K1.SAHM1 = topTab_KOPT_K1.DMSOvsKOPT_K1.SAHM1, 
INT = topTab_INT)

listOfSelected <- list()
for (i in 1:length(listOfTables)){
# select the toptable
topTab <- listOfTables[[i]]
# select the genes to be included in the analysis
whichGenes<-topTab["adj.P.Val"]<0.15
selectedIDs <- rownames(topTab)[whichGenes]
# convert the ID to Entrez
EntrezIDs<- select(hgu133plus2.db, selectedIDs, c("ENTREZID"))
EntrezIDs <- EntrezIDs$ENTREZID
listOfSelected[[i]] <- EntrezIDs
names(listOfSelected)[i] <- names(listOfTables)[i]
}

sapply(listOfSelected, length)
```
The analysis also requires to have the Entrez Identifiers for all genes analyzed. It is an open discussion if what one should use is “all genes analyzed” -that is genes that have been retained in the analysis and are part of the “topTable”- or all genes available. In this case we use the second option and define our universe to be all genes that have at least one annotation in the Gene Ontology.

```{r}
mapped_genes2GO <- mappedkeys(org.Mm.egGO)
mapped_genes2KEGG <- mappedkeys(org.Mm.egPATH)
mapped_genes <- union(mapped_genes2GO , mapped_genes2KEGG)
```
The Biological significance analysis will be applied only to the first two lists. "HPB_ALL.DMSOvsHPB_ALL.SAHM1" "KOPT_K1.DMSOvsKOPT_K1.SAHM1"

_Esto deja de funcioanrme_
```{r}
listOfData <- listOfSelected[1:2]
comparisonsNames <- names(listOfData)
universe <- mapped_genes
 
for (i in 1:length(listOfData)){
  genesIn <- listOfData[[i]]
  comparison <- comparisonsNames[i]
  enrich.result <- enrichPathway(gene = genesIn,pvalueCutoff = 0.05,
                                 readable = T,
                                 pAdjustMethod = "BH",
                                 organism = "human",
                                 universe = universe)
   
   cat("##################################")
   cat("\nComparison: ", comparison,"\n")
   print(head(enrich.result))
 
if (length(rownames(enrich.result@result)) != 0) {
   write.csv(as.data.frame(enrich.result), 
              file =paste0("./results/","ReactomePA.Results.",comparison,".csv"),
              row.names = FALSE)
   
   pdf(file=paste0("./results/","ReactomePABarplot.",comparison,".pdf"))
    print(barplot(enrich.result, showCategory = 15, font.size = 4, 
  title = paste0("Reactome Pathway Analysis for ", comparison,". Barplot")))
   dev.off()
   
  pdf(file = paste0("./results/","ReactomePAcnetplot.",comparison,".pdf"))
    print(cnetplot(enrich.result, categorySize = "geneNum", schowCategory = 15, 
         vertex.label.cex = 0.75))
  dev.off()
}
}
```
In comparison HPB_ALL.DMSOvsHPB_ALL.SAHM1, 6 enriched pathways have been found, for example:  Transcriptional Regulation by TP53.

In comparison KOPT_K1.DMSOvsKOPT_K1.SAHM1, 6 enriched pathways have been found, for example: tRNA processing.

The results obtained in the analysis of biological significance are:

- a .csv file with a summary of all the enriched pathways and the associated statistics.

- a bar plot with the best enriched pathways. Height of the bar plot is the number of genes of our analysis related with that pathway. Moreover, pathways are ordered by statistical significance.

- a plot with a network of the enriched pathways and the relation among the genes included.

This netwowrk figure shows the network produced from the genes selected in the comparison
```{r}
cnetplot(enrich.result, categorySize = "geneNum", schowCategory = 15, 
vertex.label.cex = 0.75)
```

First rows and columns for Reactome results on HPB_ALL.DMSOvsHPB_ALL.SAHM1.csv comparison:
```{r}
Tab.react <- read.csv2(file.path("./results/ReactomePA.Results.HPB_ALL.DMSOvsHPB_ALL.SAHM1.csv"), sep = ",", header = TRUE, row.names = 1)

Tab.react <- Tab.react[1:4, 1:5]
knitr::kable(Tab.react, booktabs = TRUE, caption = "First rows and columns for Reactome results on HPB_ALL.DMSOvsHPB_ALL.SAHM1.csv comparison")
```

## 6. List of files
```{r}
listOfFiles <- dir("./results/") 
knitr::kable(
listOfFiles, booktabs = TRUE,
caption = 'List of files generated in the analysis',
col.names="List_of_Files"
)
```

